api/src/main/java/org/openmrs/postgres/db/hibernate/HibernatePostgresDAO.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.postgres.db.hibernate;
+
+import org.hibernate.SessionFactory;
+import org.hibernate.Session;
+import org.openmrs.api.db.CohortDAO;
+import org.openmrs.postgres.db.PostgresDAO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Hibernate implementation of the PostgresDAO
+ *
+ * @see PostgresDAO
+ * @see org.openmrs.api.context.Context
+ * @see org.openmrs.postgres.PostgresService
+ */
+public class HibernatePostgresDAO implements PostgresDAO {
+	
+	private static final Logger log = LoggerFactory.getLogger(HibernatePostgresDAO.class);
+	
+	/**
+	 * Hibernate session factory
+	 */
+	private SessionFactory sessionFactory;
+	
+	public HibernatePostgresDAO() {
+	}
+	
+	/**
+	 * Set session factory
+	 *
+	 * @param sessionFactory
+	 */
+	public void setSessionFactory(SessionFactory sessionFactory) {
+		this.sessionFactory = sessionFactory;
+	}
+	
+	/**
+	 * @see org.openmrs.postgres.db.PostgresDAO#updateSequence()
+	 */
+	@Override
+	public void updateSequence() {
+		String dialect = sessionFactory.getProperties().getOrDefault("hibernate.dialect", "").toString();
+		if (StringUtils.containsIgnoreCase(dialect, "postgresql")) {
```

Comment:
@dkayiwa , as you can see here I am making use session to check for current database that is being used by OpenMRS via the set dialect, so **the current change will never cause an issue on MySQL. I have checked it on both MySQL and PostgreSQL and everything is working fine.**

Now, to make use of database utility approach I am not able to figure out how to get access to this hibernate session in module's activator class. This session will be required so as to pass to this utility method.

Reply:
Why I need the current change ?

I need to update postgresql sequences [here](https://github.com/openmrs/openmrs-module-referencemetadata/blob/8dc640fb4e29e7b5200e9c79027e7c23050c5095/api/src/main/java/org/openmrs/module/referencemetadata/ReferenceMetadataActivator.java#L110) , after the insertion of all the concepts in Reference Metadata Activator. If I don't update the sequence here then an error with logs : https://pastebin.com/23mQFUk1 this is raised when inserting concept source via reference application on PostgreSQL. So, I need to make a call to the service I have created in this PR here.

Reply:
How about Context.getAdministrationService().executeSQL?

Reply:
How about Context.getAdministrationService().executeSQL?

Reply:
Thanks for pointing to the AdministrativeService. I don't think we can use the executeSQL directly without doing some major changes. But what I think can be done is shift this updateSequence method to AdministrativeService . Then there won't be any requirement of a separate service as I have done here. Just a call like **Context.getAdministrationService().updatePostgreSQLsequence** sequence will do the work. I think this will be right. @dkayiwa What do you think ?

Reply:
Why can't we use executeSQL?

Reply:
If we pass the update sequence queries directly to this method then error will arise for MySQL since it does not support sequences. Now if we are to make use of this method then we will have to pass an extra argument to decide whether the sent query is of update sequence or not and then handle accordingly. If it is of update sequence type then execute it only on PostgreSQL.

Reply:
What error do you get on MySQL?

Reply:
If you execute the query "SELECT setval('person_person_id_seq', (SELECT MAX(person_id) FROM person)+1);" on MySQL then you get an error : 

**5000	Error Code: 1305. FUNCTION openmrs164.setval does not exist	0.611 sec**